Userland/Libraries/LibGL/Tex/Texture2D.cpp

```diff
@@ -24,25 +22,32 @@ void Texture2D::upload_texture_data(GLuint lod, GLint internal_format, GLsizei w
     mip.pixel_data().resize(width * height);
 
     // No pixel data was supplied leave the texture memory uninitialized.
-    if (pixels == nullptr) {
+    if (pixels == nullptr)
         return;
-    }
 
     m_internal_format = internal_format;
 
-    replace_sub_texture_data(lod, 0, 0, width, height, format, type, pixels, pixels_per_row);
+    replace_sub_texture_data(lod, 0, 0, width, height, format, type, pixels, pixels_per_row, byte_alignment);
 }
 
-void Texture2D::replace_sub_texture_data(GLuint lod, GLint xoffset, GLint yoffset, GLsizei width, GLsizei height, GLenum format, GLenum type, const GLvoid* pixels, size_t pixels_per_row)
+void Texture2D::replace_sub_texture_data(GLuint lod, GLint xoffset, GLint yoffset, GLsizei width, GLsizei height, GLenum format, GLenum type, const GLvoid* pixels, GLsizei pixels_per_row, u8 byte_alignment)
 {
     auto& mip = m_mipmaps[lod];
 
     // FIXME: We currently only support GL_UNSIGNED_BYTE pixel data
     VERIFY(type == GL_UNSIGNED_BYTE);
-    VERIFY(lod < m_mipmaps.size());
     VERIFY(xoffset >= 0 && yoffset >= 0 && xoffset + width <= mip.width() && yoffset + height <= mip.height());
+    VERIFY(pixels_per_row == 0 || pixels_per_row >= xoffset + width);
+
+    // Calculate row offset at end to fit alignment
+    int const physical_width = pixels_per_row > 0 ? pixels_per_row : width;
+    u8 const component_size_bytes = sizeof(u8);
+    u8 const component_count = (format == GL_RGBA || format == GL_BGRA) ? 4 : 3;
+    size_t const physical_width_bytes = physical_width * component_count * component_size_bytes;
+    size_t const row_remainder_bytes = (physical_width - width) * component_count * component_size_bytes
+        + (byte_alignment - physical_width_bytes % byte_alignment) % byte_alignment;
```

Comment:
The last `% byte_alignment` is not necessary. `physical_width_bytes % byte_alignment` is always less than `byte_alignment`.

Reply:
True, but `byte_alignment - physical_width_bytes % byte_alignment` can be `byte_alignment` which should result in `0` - so we still need that last modulo.

Reply:
The `(A - s % A) % A` is a common pattern to calculate remainders for alignment.

Reply:
You're right. Please discard my comment... :^(